### PR TITLE
tutorial gcloud bucket

### PR DESCRIPTION
**Pull request by MalteEbner**
## Description
Adds the tutorial with 4 steps
- Introduction, decision how to handle thumbnails.
- Configure the datasource (here: Setting gcloud bucket with IAM).
- Configuring a Lightly Dataset to use this datasource.
- Create the dataset and upload embeddings and metadata.

## Tutorial as pdf
[Create a dataset from Google Cloud Storage — lightly 1.2.1 documentation.pdf](https://github.com/lightly-ai/lightly/files/7789151/Create.a.dataset.from.Google.Cloud.Storage.lightly.1.2.1.documentation.pdf)

### CODE REVIEW

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 1]
```diff
@@ -1,6 +1,127 @@
 .. _dataset-creation-gcloud-bucket:
 
-Create dataset from google cloud bucket
+Create dataset from Google cloud bucket
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:3]: Google Cloud Storage

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 2]
```diff
@@ -1,6 +1,127 @@
...
 ---------------------------------------
 
-TODO: Add the tutorial how to create a Lightly dataset from a gcloud bucket here.
+Lightly allows you to configure a remote datasource like Google cloud bucket.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:6]: Google Cloud Storage (maybe link to https://cloud.google.com/storage)

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 3]
```diff
@@ -1,6 +1,127 @@
...
+In this guide, we will show you how to setup your Google cloud bucket,
+configure your dataset to use said bucket, and only upload metadata to Lightly.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:8]: ...to Lightly while keeping your data private

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 4]
```diff
@@ -1,6 +1,127 @@
...
+
+
+Setting up Google cloud bucket
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:11]: Google Cloud Storage

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 5]
```diff
@@ -1,6 +1,127 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+For Lightly to be able to create so-called `presigned URLs/read URLs <https://cloud.google.com/storage/docs/access-control/signed-urls>`_
+to be used for displaying your data in your browser.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:14]: merge the two sentences or make it a `:` and a bulletpoint

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 6]
```diff
@@ -1,6 +1,127 @@
...
+Let us assume the bucket is called `lightly-datalake`.
+And let us assume the folder you want to use with Lightly is located at `projects/wild-animals/`
+
+**Setting Up a Service Account in IAM**
+
+1. Write down your project ID.
+You find it in the `gcloud console <console.cloud.google.com/home/dashboard>`_ under `Project Info`.
+
+2. Navigate to your bucket in the `google cloud storage browser <https://console.cloud.google.com/storage/browser>`_
+and from there to `projects/wild-animals/`. Copy the path somewhere, in this case
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:28]: remove "somewhere". Or write "clipboard"

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 7]
```diff
@@ -1,6 +1,127 @@
...
+`lightly-datalake/projects/wild-animals`.
+
+.. figure:: ./images_gcloud_bucket/screenshot_gcloud_bucket_project.png
+    :align: center
+    :alt: Browsing a google cloud bucket.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:33]: google cloud storage bucket

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 8]
```diff
@@ -1,6 +1,127 @@
...
+    :width: 60%
+
+3. Navigate to the tab `Permissions`.
+Make sure that your Access control is Uniform.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:37]: lowercase access control, lowercase uniform

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 9]
```diff
@@ -1,6 +1,127 @@
...
+If it is not, change it to Uniform.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:38]: lowercase

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 10]
```diff
@@ -1,6 +1,127 @@
...
+
+.. figure:: ./images_gcloud_bucket/screenshot_gcloud_uniform_access.png
+    :align: center
+    :alt: Ensuring a google cloud bucket has uniform access.
+    :width: 60%
+
+
+4. Navigate to `IAM & Admin -> Roles <https://console.cloud.google.com/iam-admin/roles>`_.
+There create a new role, with the title and ID `STORAGE_READ_WRITE_LIST`.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:47]: maybe structure the todos into bullet points and remove the "Then..."

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 11]
```diff
@@ -1,6 +1,127 @@
...
+.. figure:: ./images_gcloud_bucket/screenshot_gcloud_storage_role.png
+    :align: center
+    :alt: Creating a role for accessing google cloud storage.
+    :width: 60%
+
+5. Navigate to `APIs -> Credentials <https://console.cloud.google.com/apis/credentials>`_.
+Then click on `Create Credentials`, choose `Service Account` and insert the name
+`LIGHTLY_USER_WILD_ANIMALS`.
+The description can be `service account for the Lightly API to access the wild animals dataset`.
+Then click on `Create and Continue`. Choose the Role you just created, i.e. `STORAGE_READ_WRITE_LIST`.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:62]: Also structure this in to bulletpoints

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 12]
```diff
@@ -1,6 +1,127 @@
...
+    )
+Continue to the next step `Grant users access to this service account`.
+There add yourself (in form of your mail address) to both service account roles.
+Otherwise you won't be able to change this service account.
+Then click on `Done` to create the service account.
+You can change the roles of the service account in the
+`IAM <https://console.cloud.google.com/iam-admin/iam>`.
+
+6. After creation, you can find it in the list of all service accounts.
+Click on it, then change to the tab `keys`. Click on `Add key` and create a new
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:81]: Click on the user and navigate to the "keys" tab

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 13]
```diff
@@ -1,6 +1,127 @@
...
+private key in JSON Format. It will download the corresponding key file.
+
+    .. figure:: images_gcloud_bucket/screenshot_gcloud_service_account_key_creation.png
+        :align: center
+        :alt: Google Cloud Service Account Key Creation
+        :width: 60%
+
+For Lightly to be able to create embeddings and extract metadata from your data,
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:89]: Maybe add a new heading before this section

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 14]
```diff
@@ -1,6 +1,127 @@
...
+    :alt: Configure google cloud bucket datasource in Lightly Webapp
+
+
+3. As the resource path, enter the full URI to your resource eg. `gs://lightly-datalake/projects/wild-animals`
+4. Enter the Google Project ID you just wrote down.
+5. Click on `Select Credentials File` to add the key file you downloaded in the previous step.
+5. The thumbnail suffix allows you to configure
+
+    - where your thumbnails are stored when you already have generated thumbnails in your S3 bucket
+    - where your thumbnails will be stored when you want Lightly to create thumbnails for you. For this to work, the user policy you have created must possess write permissions.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:115]: must possess -> must be granted?

**Review by japrescott — DISMISSED** [review]
looks pretty good Malte! I think we should rename something and give it a bit more structure but otherwise I like it :)

**Comment by japrescott** [comment]
here is the instructions on how to sync your bucket with your drive https://cloud.google.com/storage/docs/gsutil/commands/rsync

**Comment by MalteEbner** [comment]
> here is the instructions on how to sync your bucket with your drive https://cloud.google.com/storage/docs/gsutil/commands/rsync

Thanks, I added it to the instructions.

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 15]
```diff
@@ -1,6 +1,164 @@
...
+If it is not, change it to uniform.
+
+.. figure:: ./images_gcloud_bucket/screenshot_gcloud_uniform_access.jpg
+    :align: center
+    :alt: Ensuring a google cloud bucket has uniform access.
+    :width: 60%
+
+4. Navigate to `IAM & Admin -> Roles <https://console.cloud.google.com/iam-admin/roles>`_.
+
+- There create a new role, with the title and ID `STORAGE_READ_WRITE_LIST`.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:50]: maybe name this something lightly related. LIGHTLY_DATSET_ACCESS

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 16]
```diff
@@ -1,6 +1,164 @@
...
+
+- Click on `Create Credentials`, choose `Service Account` and insert the name
+  `LIGHTLY_USER_WILD_ANIMALS`.
+- The description can be `service account for the Lightly API to access the wild animals dataset`.
+- Click on `Create and Continue`.
+- Choose the Role you just created, i.e. `STORAGE_READ_WRITE_LIST`.
+- Add a condition with the title `BUCKET_PROJECTS_WILD_ANIMALS`
+  and insert the condition below in the Condition editor. Remember to change the bucket name
+  and path to the folder. However, you must keep the "objects" inbetween.
+  For more information, head to the `IAM conditions
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:71]: great explanation. I would add it to after the actual code snippet

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 17]
```diff
@@ -1,6 +1,164 @@
...
+
+
+3. As the resource path, enter the full URI to your resource eg. `gs://lightly-datalake/projects/wild-animals`
+4. Enter the Google Project ID you wrote down in the first step.
+5. Click on `Select Credentials File` to add the key file you downloaded in the previous step.
+6. The thumbnail suffix allows you to configure
+
+- where your thumbnails are stored when you already have generated thumbnails in your S3 bucket
+- where your thumbnails will be stored when you want Lightly to create thumbnails for you.
+  For this to work, the user policy you created must be granted write permissions.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:134]: maybe note here that using `uploads=thumbnail` would be required?

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 18]
```diff
@@ -1,6 +1,164 @@
...
+- when the thumbnail suffix is not defined/empty,
+  the Lightly Webapp will load the full image even when requesting the thumbnail.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:136]: Here is the first mention of the Lightly Webapp. I think in the docs we always call it Lightly Platform. I dont care but it should be uniform :)

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 19]
```diff
@@ -1,6 +1,164 @@
...
+
+6. Press save and ensure that at least the lights for List and Read turn green.
+If you added permissions for writing, this lights should also turn green.
+
+7. Now you should be on the dataset creation page again.
+
+Create the dataset and upload embeddings and metadata.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For creating the dataset and uploading embeddings and metadta to it, you need
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:147]: -> metadata

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 20]
```diff
@@ -1,6 +1,164 @@
...
+the :ref:`lightly-command-line-tool`. Furthermore, you need to have your data locally.
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:148]: ...your data locally on your machine

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 21]
```diff
@@ -1,6 +1,164 @@
...
+and its `rsync command <https://cloud.google.com/storage/docs/gsutil/commands/rsync>`_:
+
+.. code::
+
+    gsutil -m rsync -r /local/projects/wild-animals gs://datalake-lightly/projects/wild-animals
+
+
+Use `lightly-magic` and `lightly-upload` with the following considerations:
+
+- use `input_dir=/local/projects/wild-animals`
```
Comment (japrescott) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:159]: unclear why this is part of the considerations?

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 15 above)
Comment (philippmwirth) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:50]: I think the comma should be like this: "There, create a new role with the title ..."

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 16 above)
Comment (philippmwirth) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:71]: ^this

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 21 above)
Comment (philippmwirth) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:159]: it sounds a little strange like this, maybe just mention that the input directory should be the root directory of your dataset?

**Comment by philippmwirth** [comment]
nice work :)

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 15 above)
Comment (MalteEbner) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:50]: The role itself is not at all Lightly related, it is a role allowing read, write, list for any bucket. Alternatively, one could use the role Storage Admin, that would work as well.

As the role is neither about a specific bucket, nor about a dataset in this bucket, I would leave it like this.

Comment (MalteEbner) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:50]: @philippmwirth I changed the syntax.

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 16 above)
Comment (MalteEbner) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:71]: I moved it.

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 17 above)
Comment (MalteEbner) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:134]: I added that in the last chapter about `Create the dataset and upload embeddings and metadata`, as it is used in that step.

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 18 above)
Comment (MalteEbner) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:136]: Good point, I renamed it everywhere to `Platform`.

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 21 above)
Comment (MalteEbner) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst]: I used the same input_dir as in the gustil command above.

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 22]
```diff
@@ -1,6 +1,189 @@
...
+configure your dataset to use said bucket, and only upload metadata to Lightly
+while keeping your data private.
+
+One decision you need to make first is whether you want to use thumbnails.
+Using thumbnails makes the Lightly Platform more responsive, as not always
+the full images will be loaded.
+However, the thumbnails will be stored in you bucket and thus need storage.
+You have three options:
+
+A) You want to use thumbnails, but don't have them yet. Then you need to give
```
Comment (philippmwirth) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:18]: make this a bullet list

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 23]
```diff
@@ -1,6 +1,189 @@
...
+
+3. As the resource path, enter the full URI to your resource eg. `gs://lightly-datalake/projects/wild-animals`
+4. Enter the Google Project ID you wrote down in the first step.
+5. Click on `Select Credentials File` to add the key file you downloaded in the previous step.
+6. The thumbnail suffix depends on the option you chose in the first step
+
+- Option A: You want Lightly to create the thumbnail for you.
+  Then choose the naming scheme to your liking.
+- Option B: You have already generated thumbnails in your bucket.
+  Then choose the thumbnail suffix such that it reflects you naming scheme.
```
Comment (philippmwirth) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:159]: *your naming scheme

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst [hunk 24]
```diff
@@ -1,6 +1,189 @@
...
+- Option C: You don't want to use thumbnails.
+  Then leave the thumbnail suffix undefined/empty.
+
+
+
+6. Press save and ensure that at least the lights for List and Read turn green.
+If you added permissions for writing to use option A, this light should also turn green.
+
+7. Now you should be on the dataset creation page again.
```
Comment (philippmwirth) [inline docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:168]: That's not true, you are still on the `edit dataset` screen. I would just leave this bullet point away

Comment (philippmwirth) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:168]: Or write that you can now exit the settings by clicking the X

**Review by philippmwirth — CHANGES_REQUESTED** [review]
Three more tiny changes then we're good to go :)

File: docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst (hunk 24 above)
Comment (MalteEbner) [reply docs/source/getting_started/dataset_creation/dataset_creation_gcloud_bucket.rst:168]: Good point, I changed it to `7. After closing the pop-up by clicking the X, you should be on the dataset creation page again.`.

**Comment by MalteEbner** [comment]
I also updated the pdf.

**Review by philippmwirth — APPROVED** [review]

**Pull request merged**